### ISSUE TITLE
Api always returns 401 after Amplify Update

### ISSUE THREAD

**Issue by iThink32** [aws-amplify/amplify-ios#1439]
### Describe the bug

i am facing an issue in which i have an iOS project setup which uses cognito to authenticate users , it logs in successfully but all api's return 401.

I have followed all 11 steps [here](https://github.com/aws-amplify/amplify-cli/issues/3390)
 and the concerning part is this is happening after i updated amplify to the following

Installing AWSAuthCore 2.25.0 (was 2.14.2)
Installing AWSCognitoIdentityProvider 2.25.0 (was 2.14.2)
Installing AWSCognitoIdentityProviderASF 2.25.0 (was 1.0.1)
Installing AWSCore 2.25.0 (was 2.14.2)
Installing AWSMobileClient 2.25.0 (was 2.14.2)
Installing AWSPluginsCore 1.13.4 (was 1.0.4)
Installing AWSS3 2.25.0 (was 2.14.2)
Installing Amplify 1.13.4 (was 1.0.4)
Installing AmplifyPlugins 1.13.4 (was 1.0.4)
Installing AppSyncRealTimeClient 1.5.0 (was 1.1.6)
The earlier version is also provided above.

If i switch back to the earlier version of amplify plugins it works fine.Can you pls help me out?

### Steps To Reproduce

```swift
Switch from an older version of amplify mentioned above to the new one also mentioned above
```

### Expected behavior

It should work as it is working in older versions

### Amplify Framework Version

2.25.0 

### Amplify Categories

API, Auth, Storage

### Dependency manager

Cocoapods

### Swift version

5.1

### CLI version

6.0.1

### Xcode version

12.5.1

### Relevant log output

```shell
Api always returns 401 after Amplify Update
```

### Is this a regression? (i.e. was this working before a version upgrade)

its not a regression but yes it is working on an older version

### Device

iPhone8

### iOS Version

iOS 14

**Comment by iThink32** [comment]
@lawmicha pls look into this as your help would be highly appreciative.

**Comment by lawmicha** [comment]
Hi @iThink32, that link you provided https://github.com/aws-amplify/amplify-cli/issues/3390 is a feature request I opened on the CLI and is to help describe the current experience without the feature, so that is to say if it is open then the use case is not supported out of the box and the steps will get out of date. 

Could you let us know what is your use case?  Which plugins are you using? API and Storage? Which services are you using? APIGateway and S3?

How do you expect the APIs to be authenticated?

APIGateway, as followed in those steps, will accept ID token or Access Token. APIPlugin would need to add the correct token. You can also set the authorizationType to None and inject the token yourself https://docs.amplify.aws/lib/restapi/authz/q/platform/ios/#note-related-to-use-access-token-or-id-token

AWSS3StoragePlugin retrieves the Cognito Identity Id and adds its on for private and protected access levels, and does IAM signing. S3 supports user pool token authentication.

**Comment by iThink32** [comment]
@lawmicha to give you additional info , it is related to linking cognito to Storage , for api gateway i had to add an auth scope to make it work due to an escaping issue in the way the Authentication Header is sent from iOS as seen above or aws-amplify/amplify-flutter#471.
If you could tell me how do i do the same for Storage it would help me solve this problem.
Moreover the same access works fine when cli IAM controls are used but not when auth is set to COGNITO_USER_POOLS

**Comment by iThink32** [comment]
@lawmicha i will try the above solution provided i.e NONE and get back to you but to answer your questions

Could you let us know what is your use case? 
I am using cognito to sign in users post which i use storage to upload / download files which i want only signed in / authenticated users to access.But as of now it always throws a 403 irrespective of whether im accessing the public or protected bucket and even for authenticated users.

Which plugins are you using? 
Auth
Storage
APIGateway
```
            try Amplify.add(plugin: AWSCognitoAuthPlugin())
            try Amplify.add(plugin: AWSS3StoragePlugin())
            try Amplify.add(plugin: AWSAPIPlugin())
```

Which services are you using? 
APIGateway , Cogntio and S3 

How do you expect the APIs to be authenticated?
Cognito only

**Comment by iThink32** [comment]
@lawmicha any update on this? , i agree that the approach you suggested can be used for APIGateway to manually add headers but what about upload/ download , there is no option to add headers manually there

**Comment by iThink32** [comment]
@lawmicha i would request you to pls assign this issue to yourself.
Frankly speaking its been two months and im thinking of considering other options due to the lack to help ive received on all mediums of communication.

**Comment by iThink32** [comment]
@lawmicha thanks for your response but ive tried what you told me earlier but to no luck.
Fortunately i've made it work by the following solution and i will also reason it out.Please correct me if wrong.

In the project's amplifyconfiguration, remove the following line
```
"defaultAccessLevel": "guest"
```
The above line is a part of the storage configuration and will finally look like this

```
    "storage": {
        "plugins": {
            "awsS3StoragePlugin": {
                "bucket": “yourBucket”,
                "region": “yourRegion
            }
        }
    }
    ```
    
    After this use download and upload using the protected access specifier like 
    
    ```
    let options = StorageUploadDataRequest.Options(accessLevel: .protected,
                                                       targetIdentityId: identityID)
        let uploadOperationTask = Amplify.Storage.uploadData(key: filename,data: data,
                                                             options: options ...
                                                        
 ```                                                     
 
 Reason:
 The line i has requested to remove to make it work is because it assumes only that role and nothing else can change it even if authenticated.Since the guest role uses the unauth role it will always throw 401 or 403.

**Comment by iThink32** [comment]
@lawmicha the above fix does not work on the lastest version of CLI released a day back i.e 7.4.4

I set the Auth Type to NONE and injected the Authorization Header myself and all api's seem to work fine but storage returns the following

```
StorageError: Unexpected error occurred with message: Domain: [Amplify.AuthError
Code: [2
LocalizedDescription: [The operation couldn’t be completed. (Amplify.AuthError error 2.)
LocalizedFailureReason: [
LocalizedRecoverySuggestion: [
Recovery suggestion: This should not happen. There is a possibility that there is a bug if this error persists. Please take a look at https://github.com/aws-amplify/amplify-ios/issues to see if there are any existing issues that match your scenario, and file an issue with the details of the bug if there isn't. Issue encountered at:
file: ...StorageError.swift
function: recoverySuggestion
line: 63
StorageError: Unexpected error occurred with message: Domain: [Amplify.AuthError
Code: [2
LocalizedDescription: [The operation couldn’t be completed. (Amplify.AuthError error 2.)
LocalizedFailureReason: [
LocalizedRecoverySuggestion: [
Recovery suggestion: This should not happen. There is a possibility that there is a bug if this error persists. Please take a look at https://github.com/aws-amplify/amplify-ios/issues to see if there are any existing issues that match your scenario, and file an issue with the details of the bug if there isn't. Issue encountered at:
file: ...StorageError.swift
function: recoverySuggestion
line: 63
```
Can you pls suggest a direction for the same?

**Comment by lawmicha** [comment]
Hi @iThink32, thanks for providing the code snippet and details of your issue regarding `defaultAccessLevel`. 

`"defaultAccessLevel": "guest"` is a configuration value that allows you to specific which access level is used in the Amplify.Storage request when no `accessLevel` is passed in. Removing this while it defaults to `guest` does not do anything to the requests (since it will always default to guest). The use case of this configuration value allows you to set it to something else like `"defaultAccessLevel": "protected"` then your Storage requests will always use the protected access level. By doing this:
```
let options = StorageUploadDataRequest.Options(accessLevel: .protected,
                                                       targetIdentityId: identityID)
```

You're explicitly setting the access level to use protected anyways. so keeping `"defaultAccessLevel": "guest"` will have no impact to your app. You can verify this by adding the `defaultAccessLevel` back in and making the request with accessLevel protected and should continue to work as you have.

> The line i has requested to remove to make it work is because it assumes only that role and nothing else can change it even if authenticated.Since the guest role uses the unauth role it will always throw 401 or 403.

Sorry I didn't quite follow this. When you provision your S3 bucket, you can decide what type of access your users have
```
? What kind of access do you want for Authenticated users? create/update, read, delete
? What kind of access do you want for Guest users? create/update, read, delete
```
You can read more about the file access levels here https://docs.amplify.aws/lib/storage/configureaccess/q/platform/ios/

**Comment by lawmicha** [comment]
> @lawmicha the above fix does not work on the lastest version of CLI released a day back i.e 7.4.4

Can you expand on how this is relevant with the latest version of the CLI? What steps did you do in the CLI to provision your backend?

> 
> I set the Auth Type to NONE and injected the Authorization Header myself and all api's seem to work fine but storage returns the following
> 
> ```
> StorageError: Unexpected error occurred with message: Domain: [Amplify.AuthError

Without more details, we won't be able to debug the issue you are seeing. Would it be possible to provide us with the `amplifyconfiguration.json` file with the values redacted so that we can see the structure of the configuration, without leaking the actual values here.

What is the calling code for the Storage request? 

Do you have a stack trace when you get this error? you can put a breakpoint and view the stack to see which methods are being called. You can copy this and paste it over here as well.

**Comment by iThink32** [comment]
@lawmicha _Can you expand on how this is relevant with the latest version of the CLI? What steps did you do in the CLI to provision your backend?_

- i removed the defaultAccessLevel line from amplifyConfiguration.json in the previous version and used the protectedAccessLevel in both upload and download and it worked fine.

_Without more details, we won't be able to debug the issue you are seeing. Would it be possible to provide us with the amplifyconfiguration.json file with the values redacted so that we can see the structure of the configuration, without leaking the actual values here._

Sure , here is the file

```
{
    "UserAgent": "aws-amplify-cli/2.0",
    "Version": "1.0",
    "api": {
        "plugins": {
            "awsAPIPlugin": {
                "APIOne": {
                    "endpointType": "REST",
                    "endpoint": "https://endpoint",
                    "region": "region",
                    "authorizationType": "NONE"
                },
                .... other apis
            }
        }
    },
    "auth": {
        "plugins": {
            "awsCognitoAuthPlugin": {
                "UserAgent": "aws-amplify/cli",
                "Version": "0.1.0",
                "IdentityManager": {
                    "Default": {}
                },
                "CredentialsProvider": {
                    "CognitoIdentity": {
                        "Default": {
                            "PoolId": "PoolID",
                            "Region": "Region"
                        }
                    }
                },
                "CognitoUserPool": {
                    "Default": {
                        "PoolId": "PoodID",
                        "AppClientId": "AppClientID",
                        "Region": "Region"
                    }
                },
                "Auth": {
                    "Default": {
                        "authenticationFlowType": "USER_SRP_AUTH",
                        "loginMechanisms": [
                            "EMAIL",
                            "PHONE_NUMBER"
                        ],
                        "signupAttributes": [
                            "EMAIL"
                        ],
                        "passwordProtectionSettings": {
                            "passwordPolicyMinLength": 8,
                            "passwordPolicyCharacters": []
                        },
                        "mfaConfiguration": "ON",
                        "mfaTypes": [
                            "SMS"
                        ],
                        "verificationMechanisms": [
                            "EMAIL"
                        ]
                    }
                },
                "S3TransferUtility": {
                    "Default": {
                        "Bucket": "BucketName",
                        "Region": "Region"
                    }
                }
            }
        }
    },
    "storage": {
        "plugins": {
            "awsS3StoragePlugin": {
                "bucket": "BucketName",
                "region": "Region",
                "defaultAccessLevel": "guest"
            }
        }
    }
}
``` 

_What is the calling code for the Storage request?_

```
        let options = StorageUploadDataRequest.Options(accessLevel: .protected,
                                                       targetIdentityId: identityID)
        let uploadOperationTask = Amplify.Storage.uploadData(key: filename,
                                                             data: data,
                                                             options: options,
                                       progressListener: { (progress) in
            progressHandler?(progress)
        }, resultListener: { (result) in
            switch result {
            case Result.success :
                completion(nil)
            case let Result.failure(error):
                completion(NetworkingError.apiError(description: error.errorDescription))
                print(error)
            }
        })
        return UploadTask(task: uploadOperationTask)
 ```
 
 _Do you have a stack trace when you get this error? you can put a breakpoint and view the stack to see which methods are being called. You can copy this and paste it over here as well._
 
 A point to be noted , it at times works but if i try it twice or thrice i get the following.Please note this is the complete log after trying three times and failing each time
 
 ```
StorageError: Unexpected error occurred with message: Domain: [Amplify.AuthError
Code: [2
LocalizedDescription: [The operation couldn’t be completed. (Amplify.AuthError error 2.)
LocalizedFailureReason: [
LocalizedRecoverySuggestion: [
Recovery suggestion: This should not happen. There is a possibility that there is a bug if this error persists. Please take a look at https://github.com/aws-amplify/amplify-ios/issues to see if there are any existing issues that match your scenario, and file an issue with the details of the bug if there isn't. Issue encountered at:
file: /Users/nashashank/Documents/brent/Brent-iOS/Pods/Amplify/Amplify/Categories/Storage/Error/StorageError.swift
function: recoverySuggestion
line: 63

StorageError: Unexpected error occurred with message: Domain: [Amplify.AuthError
Code: [2
LocalizedDescription: [The operation couldn’t be completed. (Amplify.AuthError error 2.)
LocalizedFailureReason: [
LocalizedRecoverySuggestion: [
Recovery suggestion: This should not happen. There is a possibility that there is a bug if this error persists. Please take a look at https://github.com/aws-amplify/amplify-ios/issues to see if there are any existing issues that match your scenario, and file an issue with the details of the bug if there isn't. Issue encountered at:
file: /Users/nashashank/Documents/brent/Brent-iOS/Pods/Amplify/Amplify/Categories/Storage/Error/StorageError.swift
function: recoverySuggestion
line: 63

StorageError: Unexpected error occurred with message: Could not retreive AWS credentials, credential value is nil or empty.
Recovery suggestion: This should not happen. There is a possibility that there is a bug if this error persists. Please take a look at https://github.com/aws-amplify/amplify-ios/issues to see if there are any existing issues that match your scenario, and file an issue with the details of the bug if there isn't. Issue encountered at:
file: /Users/nashashank/Documents/brent/Brent-iOS/Pods/Amplify/Amplify/Categories/Auth/Error/AuthError.swift
function: recoverySuggestion
line: 80

StorageError: Unexpected error occurred with message: Could not retreive AWS credentials, credential value is nil or empty.
Recovery suggestion: This should not happen. There is a possibility that there is a bug if this error persists. Please take a look at https://github.com/aws-amplify/amplify-ios/issues to see if there are any existing issues that match your scenario, and file an issue with the details of the bug if there isn't. Issue encountered at:
file: /Users/nashashank/Documents/brent/Brent-iOS/Pods/Amplify/Amplify/Categories/Auth/Error/AuthError.swift
function: recoverySuggestion
line: 80
```

 Stack Trace:
 
<img width="480" alt="Screenshot 2021-11-23 at 9 32 49 PM" src="https://user-images.githubusercontent.com/34152717/143061090-6e2333fd-5e8e-43f9-8017-8840fc910a7f.png">

 <img width="478" alt="Screenshot 2021-11-23 at 9 32 35 PM" src="https://user-images.githubusercontent.com/34152717/143061058-21ea5fd0-9b0b-40f5-8f99-435c03dd3a15.png">

**Comment by lawmicha** [comment]
thanks for the details @iThink32
> it at times works but if i try it twice or thrice i get the following.Please note this is the complete log after trying three times and failing each time

If it works sometimes, this indicates that your backend is set up correctly with all of the IAM policies and authorization rules in place.  The logs
```
Could not retreive AWS credentials, credential value is nil or empty.
```

indicate that this may be related to a similar issue over in https://github.com/aws-amplify/amplify-ios/issues/1501 

We'll keep you posted as we investigate further

**Comment by iThink32** [comment]
sure @lawmicha

**Comment by brennanMKE** [comment]
@iThink32 Please update to the latest release for Amplify and try it with your code and let us know if is now working for you.

**Comment by brennanMKE** [comment]
You will see we have a fix in for Auth in a [recent release](https://github.com/aws-amplify/amplify-ios/blob/main/CHANGELOG.md).

**Comment by iThink32** [comment]
@brennanMKE  awesome. Cool will try it out and let you know

**Comment by iThink32** [comment]
@brennanMKE updated to the latest version and using the upload download feature from
a few days , as of now it has failed only once, would test it for a few more days and confirm the same but in a gist huge improvement and so far so good

**Comment by brennanMKE** [comment]
@iThink32 That's a great improvement. We have more improvements on the way. When you do experience an issue please do not hesitate to provide the details in a new issue.

**Comment by iThink32** [comment]
@brennanMKE guys i've discovered two more issues on using s3 a lot in the couple of days
1. The progress object only returns 0 or 100 
i'm using the following callback 
```
Amplify.Storage.uploadData(key: filename,
                                                             data: data,
                                                             options: options,
                                       progressListener: { (progress) in
            progressHandler?(progress)
```

2. Since yesterday s3 has become very very slow

**Comment by iThink32** [comment]
@brennanMKE if you want i can create a new issue for the same

**Comment by brennanMKE** [comment]
@iThink32 Please create new issues so we can triage them.

**Comment by iThink32** [comment]
@brennanMKE  https://github.com/aws-amplify/amplify-ios/issues/1621 pls check this , all details specified there
